kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,406 @@
+                    Extensions(ExpExtension::ExtendedWitness(i))
+                }
+                ExtendedFoldingColumn::Error => Extensions(ExpExtension::Error),
+                ExtendedFoldingColumn::Shift => Extensions(ExpExtension::Shift),
+                ExtendedFoldingColumn::UnnormalizedLagrangeBasis(i) => UnnormalizedLagrangeBasis(i),
+                ExtendedFoldingColumn::Constant(c) => Constant(c),
+                ExtendedFoldingColumn::Challenge(c) => Challenge(c),
+                ExtendedFoldingColumn::Alpha(i) => Extensions(ExpExtension::Alpha(i)),
+            },
+            FoldingExp::Double(exp) => Double(Box::new(exp.into_compatible())),
+            FoldingExp::Square(exp) => Square(Box::new(exp.into_compatible())),
+            FoldingExp::Add(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Add, e1, e2)
+            }
+            FoldingExp::Sub(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Sub, e1, e2)
+            }
+            FoldingExp::Mul(e1, e2) => {
+                let e1 = Box::new(e1.into_compatible());
+                let e2 = Box::new(e2.into_compatible());
+                BinOp(Op2::Mul, e1, e2)
+            }
+        }
+    }
+}
+impl std::ops::Add for Degree {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) | (One, One) => One,
+            (_, Two) | (Two, _) => Two,
+        }
+    }
+}
+impl std::ops::Mul for &Degree {
+    type Output = Degree;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        use Degree::*;
+        match (self, rhs) {
+            (Zero, Zero) => Zero,
+            (Zero, One) | (One, Zero) => One,
+            (Zero, Two) | (One, One) | (Two, Zero) => Two,
+            (One, Two) | (Two, One) | (Two, Two) => panic!("degree over 2"),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Term<C: FoldingConfig> {
+    pub exp: FoldingExp<C>,
+    pub sign: bool,
+}
+
+impl<C: FoldingConfig> Term<C> {
+    /*fn map_exp<F>(self, f: F) -> Self
+    where
+        F: Fn(FoldingExp2<C>) -> FoldingExp2<C>,
+    {
+        let Self { exp, sign } = self;
+        let exp = f(exp);
+        Self { exp, sign }
+    }*/
+    fn wrap_exp<F>(self, f: F) -> Self
+    where
+        F: Fn(Box<FoldingExp<C>>) -> FoldingExp<C>,
+    {
+        let Self { exp, sign } = self;
+        let exp = f(Box::new(exp));
+        Self { exp, sign }
+    }
+}
+impl<C: FoldingConfig> std::ops::Mul for &Term<C> {
+    type Output = Term<C>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        #[allow(clippy::suspicious_arithmetic_impl)]
+        let sign = !(self.sign ^ rhs.sign);
```

Comment:
Clippy is warning here for a different reason, but this also strikes me as suspicious/confusing.